test/flush_worker_test.cc

```diff
@@ -0,0 +1,130 @@
+    EXPECT_CALL(*timer_, enableTimer(_, _))
+        .WillRepeatedly(Invoke([&](const std::chrono::microseconds,
+                                   const Envoy::ScopeTrackedObject*) { timer_set_ = true; }));
+    EXPECT_CALL(*timer_, disableTimer()).WillOnce(Invoke([&]() { timer_set_ = false; }));
+    EXPECT_CALL(*dispatcher_, exit()).WillOnce(Invoke([&]() { stopped_ = true; }));
+  }
+
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  void simulateTimerLoop() {
+    while (!stopped_) {
+      if (timer_set_) {
+        timer_set_ = false;
+        timer_cb_();
+      }
+    }
+  }
+
+  NiceMock<Envoy::Api::MockApi> mock_api_;
+  Envoy::Thread::ThreadFactory& thread_factory_;
+  Envoy::Stats::IsolatedStoreImpl store_;
+  NiceMock<Envoy::ThreadLocal::MockInstance> tls_;
+  Envoy::Random::RandomGeneratorImpl rand_;
+  NiceMock<Envoy::Event::MockDispatcher>* dispatcher_ = nullptr;
+  std::unique_ptr<Envoy::Runtime::ScopedLoaderSingleton> loader_;
+  NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Envoy::ProtobufMessage::MockValidationVisitor> validation_visitor_;
+
+  NiceMock<Envoy::Event::MockTimer>* timer_; // not owned
+  Envoy::Event::TimerCb timer_cb_;
+  bool timer_set_{};
+  bool stopped_{};
+
+  Envoy::Stats::MockSink* sink_ = nullptr;
+  std::list<std::unique_ptr<Envoy::Stats::Sink>> stats_sinks_;
+  std::chrono::milliseconds stats_flush_interval_{10};
+};
+
+TEST_F(FlushWorkerTest, WorkerFlushStatsPeriodically) {
+  setupDispatcherTimerEmulation();
+
+  auto worker = std::make_unique<FlushWorkerImpl>(mock_api_, tls_, store_, stats_sinks_,
+                                                  stats_flush_interval_);
+
+  std::thread thread = std::thread([&worker] {
+    sleep(1);
+    worker->exitDispatcher();
+  });
+
+  expectDispatcherRun();
+  // During the 1s sleep, timer_cb_ in simulateTimerLoop() should be called at
+  // least 100 times.
+  EXPECT_CALL(*sink_, flush(_)).Times(testing::AtLeast(100));
+
+  worker->start();
+  worker->waitForCompletion();
+  thread.join();
+  worker->shutdown();
+}
+
+TEST_F(FlushWorkerTest, FinalFlush) {
+  auto worker = std::make_unique<FlushWorkerImpl>(mock_api_, tls_, store_, stats_sinks_,
+                                                  stats_flush_interval_);
+
+  // Final flush should be done in FlushWorkerImpl::shutdownThread().
+  EXPECT_CALL(*sink_, flush(_)).Times(1);
+
+  worker->start();
+  worker->waitForCompletion();
+  worker->shutdown();
```

Comment:
does this guarantee that the flush we observed is called after `FlushWorkerImpl::shutdownThread()`? I think that's what the crown jewel is that we are after here right? Catching that particular even is important, because if that breaks, my concern is that this will slip by unnoticed during usage if it breaks in the future.

Reply:
Yes, this will verify the only flush done is the final flush in `FlushWorkerImpl::shutdownThread()` (note according to current implementation, the final flush is done inside `shutdownThread()` not after).

Added comment to clarify this.

Reply:
I see, but what's not immediately clear to me is how it is guaranteed that the flush happens after the `shutdownThread()` call, and not before it? (I might be a little dense here, but I bet that someone else might have the same question pop up while reading this, so possibly a code-level comment explaining why this is guaranteed would be helpful here?).

Reply:
Ah, the final flush call `flushStats()` happens inside `FlushWorkerImpl::shutdownThread()` as you can see in the flush_worker_impl.cc. And the `FlushWorkerImpl::shutdownThread()` is executed when `flush_worker_->shutdown()` is called in `ProcessImpl::shutdown()` in process_impl.cc.

Also added comment in the code.

Reply:
Ah, so this works on the premise that if there's a single flush, it can only be the final one induced via `worker->shutdown();`, correct?
Would it then be better to move the expectation to right before `worker->shutdown();` to make sure nothing in the lines that executed before it triggered the flush?

Reply:
Good point. Moved to right before `worker->shutdown()`;